cache/src/test/kotlin/com/dropbox/android/external/cache4/CacheLoaderTest.kt

```diff
@@ -0,0 +1,314 @@
+                cache.put(1, "cat")
+
+                // invalidate the entry
+                cache.invalidate(1)
+            }
+
+            delay(executionTime + 10)
+
+            verify(loader, times(1)).invoke()
+
+            // entry from loader should be cached as previous one had been invalidated.
+            assertThat(value)
+                .isEqualTo("dog")
+
+            assertThat(cache.get(1))
+                .isEqualTo("dog")
+        }
+
+    @Test
+    fun `only 1 loader is executed by multiple concurrent get(key, loader) calls`() =
+        runBlocking {
+            val cache = CacheBuilder()
+                .build<Long, String>()
+
+            val executionTime = 20L
+
+            val loader = createSlowLoader("cat", executionTime)
+
+            repeat(3) {
+                launch(Dispatchers.Default) {
+                    cache.get(1, loader)
+                }
+            }
+
+            delay(executionTime * 3 + 10)
+
+            verify(loader, times(1)).invoke()
+
+            assertThat(cache.get(1))
+                .isEqualTo("cat")
+        }
+
+    @Test
+    fun `a loader exception is propagated to the get(key, loader) call`() {
+        val cache = CacheBuilder()
+            .build<Long, String>()
+
+        val loader = createFailingLoader<String>(IOException())
+
+        assertThrows(IOException::class.java) {
+            cache.get(1, loader)
+        }
+
+        verify(loader, times(1)).invoke()
+
+        assertThat(cache.get(1))
+            .isNull()
+    }
+
+    @Test
+    fun `a blocked concurrent get(key, loader) call is unblocked and executes its own loader after the loader from an earlier concurrent call throws an exception`() =
+        runBlocking {
+            val cache = CacheBuilder()
+                .build<Long, String>()
+
+            val executionTime = 50L
+
+            val loader1 = createSlowFailingLoader<String>(IOException(), executionTime)
+            val loader2 = createLoader("cat")
+
+            var value: String? = null
+
+            launch(Dispatchers.Default) {
+                runCatching {
+                    cache.get(1, loader1)
+                }
+            }
+
+            launch(Dispatchers.Default) {
+                delay(1)
+                value = cache.get(1, loader2)
+            }
+
+            delay(executionTime * 2)
+
+            verify(loader1, times(1)).invoke()
+            verify(loader2, times(1)).invoke()
+
+            assertThat(value)
+                .isEqualTo("cat")
+
+            assertThat(cache.get(1))
+                .isEqualTo("cat")
+        }
+
+    private fun <Value> createLoader(computedValue: Value): () -> Value {
+        return spy<() -> Value>().apply {
+            whenever(invoke()).thenReturn(computedValue)
+        }
+    }
+
+    private fun <Value> createSlowLoader(
+        computedValue: Value,
+        executionTime: Long
+    ): () -> Value {
+        return spy<() -> Value>().apply {
```

Comment:
Can we avoid using mocks and spys here all you need is a simple invocation counter:
```
class TestLoader<T>(private val block: () -> T) {
  var invokeCount = 0
  operation fun invoke(): T {
    invokeCount++
    return block() 
  }
```